design/design-doc-uptest-improvements-and-increasing-test-coverage.md

```diff
@@ -0,0 +1,344 @@
+test locally.
+
+By default, `Secret` source is still used, but a specific provider config
+manifest can be applied to the cluster via a CLI flag:
+
+```shell
+uptest e2e --provider-config="examples/irsa-config.yaml" --data-source="${UPTEST_DATASOURCE_PATH}" --setup-script=cluster/test/setup.sh --default-conditions="Test"
+```
+
+### More Comprehensive Test Assertions
+
+`Uptest` focuses on the status conditions of the crossplane resources. For
+example, during a test of MR, the `Uptest` checks the `UpToDate` condition and
+does not look to the fields of the created resources. Doing more comprehensive
+assertions like comparing the values of the fields in spec and status of MRs and
+validating patch steps for Compositions will increase the test coverage.
+
+Comparisons can be made here using Crossplane's `fieldpath` library. The set of
+fields in `status.AtProvider` has, with recent changes, become a set that
+includes those in `spec.ForProvider`. In this context, comparisons can be made
+using a go tool written using the capabilities of the `fieldpath` library.
+
+```go
+// ...
+
+pv, err := fieldpath.PaveObject(mg)
+if err != nil {
+    return nil, errors.Wrap(err, "cannot pave the managed resource")
+}
+
+specV, err := pv.GetValue("spec.forProvider")
+if err != nil {
+    return nil, errors.Wrap(err, "cannot get spec.forProvider value from paved object")
+}
+specM, ok := specV.(map[string]any)
+if !ok {
+    return nil, errors.Wrap(err, "spec.forProvider must be a map")
+}
+
+statusV, err := pv.GetValue("status.atProvider")
+if err != nil {
+return nil, errors.Wrap(err, "cannot get status.atProvider value from paved object")
+}
+statusM, ok := statusV.(map[string]any)
+if !ok {
+return nil, errors.Wrap(err, "status.atProvider must be a map")
+}
+
+for key, value := range specM {
+    // Recursively compare the spec fields with status fields
+	// ...
+}
+
+// ...
+```
+
+Related Issue: https://github.com/upbound/official-providers-ci/issues/175
+
+### Mocking Providers
+
+`Uptest` provisions physical resources while running tests on providers. In some
+cases, users may want to run their tests on a mock system. Mocking providers is
+not directly the subject of `Uptest`, but enabling some existing mock
+infrastructures is the point of `Uptest`.
+
+For example, [localstack](https://github.com/localstack/localstack) is a cloud
+service emulator that runs in a single container on your laptop or in your CI
+environment. With LocalStack, you can run your AWS applications or Lambdas
+entirely on your local machine without connecting to a remote cloud provider.
+
+Currently, there is an infrastructure for localstack in `provider-aws`. If
+`ProviderConfig` is configured properly, it will be possible to perform the
+relevant tests in a mock way. Therefore, increasing the `ProviderConfig`
+coverage mentioned before and even allowing custom configurations is at a
+critical point in this context.
```

Comment:
```suggestion
Currently, there is an ability to use LocalStack in `provider-aws`. If
`ProviderConfig` is configured properly, it will be possible to perform the
relevant tests in a mocked way. Therefore, increasing the `ProviderConfig`
coverage mentioned before and even allowing custom configurations is key
to unlocking this capability.
```

Reply:
Done!